hack/run-hive-locally.sh

```diff
@@ -0,0 +1,365 @@
+    # Blocks until the new hive-clustersync is created and has zero replica
+    while ! oc wait -n "$hive_controllers_ns" --for=jsonpath='{.spec.replicas}'=0 sts/"$hive_component_clustersync"; do
+        sleep 1
+    done
+}
+
+function disable_maintenance_mode {
+    hl_echo "Disabling maintenance mode"
+    oc patch hiveconfig hive -p '{"spec": {"maintenanceMode": false}}' --type merge
+
+    hl_echo "Waiting for the controllers to be scaled up"
+    # Blocks until the existing hive-controllers is updated to have 1 replica
+    while ! oc wait -n "$hive_controllers_ns" --for=jsonpath='{.status.updatedReplicas}'=1 deploy/"$hive_component_controllers"; do
+        sleep 1
+    done
+    # Blocks until the new hive-clustersync is created and has the correct number of replicas
+    while ! oc wait -n "$hive_controllers_ns" --for=jsonpath='{.status.updatedReplicas}'="$hive_clustersync_replicas" sts/"$hive_component_clustersync"; do
+        sleep 1
+    done
+}
+
+function run_hive_operator_locally {
+    hl_echo "Scaling down Deployment/hive-operator"
+    oc scale -n "$hive_operator_ns" deploy/"$hive_component_operator" --replicas=0
+    oc wait -n "$hive_operator_ns" --for=jsonpath='{.spec.replicas}'=0 deploy/"$hive_component_operator"
+
+    hl_echo "Running $component locally, press ctrl + c when you are done"
+    local final_env_values=(
+        HIVE_OPERATOR_NS="$hive_operator_ns" \
+        LOG_LEVEL="$hive_operator_log_level"
+    )
+    env "${final_env_values[@]}" make run-operator 2>&1 | tee "./$artifacts_folder/${hive_component_operator}-logs.txt"
+}
+
+function run_hive_controllers_locally {
+    while true; do
+        # Declare (global) associative arrays
+        declare -gA resolved_envs resolved_envFrom_cms
+
+        # Get in-cluster configurations
+        get_incluster_config
+        resolved_envs["HIVE_NS"]="HIVE_NS=$hive_controllers_ns"
+        resolved_envs["HIVE_CLUSTERSYNC_POD_NAME"]="HIVE_CLUSTERSYNC_POD_NAME=hive-clustersync-0"
+
+        # Enable maintenance mode and wait for the controllers to be scaled down
+        enable_maintenance_mode
+
+        # Run hive-controllers in the background
+        hl_echo "Running $hive_component_controllers locally, press ctrl + c when you are done"
+        # Build the manager binary for the first iteration only
+        if [ $hc_spec_generation -eq 0 ]; then
+            env "${resolved_envs[@]}" LOG_LEVEL="$hive_controllers_log_level" make run 2>&1 |
+            tee "./$artifacts_folder/${hive_component_controllers}-logs-${hc_spec_generation}.txt" &
+        else
+            env "${resolved_envs[@]}" ./bin/manager --log-level="$hive_controllers_log_level" 2>&1 |
+            tee "./$artifacts_folder/${hive_component_controllers}-logs-${hc_spec_generation}.txt" &
+        fi
+
+        # Re-deploy the controllers once HiveConfig.spec changes
+        wait_until_hiveconfig_spec_changes
```

Comment:
This seems odd. If hiveconfig hasn't changed recently, won't this hang indefinitely?

Reply:
This is meant to hang until HiveConfig.spec changes. As a result, HiveConfig.spec changes => the program proceed to the next iteration of the while loop => get the latest in-cluster configurations => run controllers locally => repeat.

Reply:
I get it now -- it was the `&`s on L238/241 that made it click for me.

So you don't have any control over the controllers getting restarted if something about the hiveconfig changes. Could that disrupt a debugging session? (I guess just... don't change the hiveconfig :P)

Reply:
I guess it won't, as this aligns with the in-cluster functioning of Hive -- when HiveConfig is updated, (most often) Hive components are re-deployed to reflect the configuration change. And without rebuilding the binaries, local restarts are just slightly slower than the in-cluster re-deployment. 

But yeah, we could add a flag which controls this behavior -- whether to restart on HiveConfig.spec changes or not.